src/completion.mts

```diff
@@ -6,122 +6,280 @@ import {
-/** https://tc39.es/ecma262/#sec-completion-record-specification-type */
-export @callable((_CompletionClass, _thisValue, [init]) => {
+let createNormalCompletion: <T>(init: NormalCompletionInit<T>) => NormalCompletionImpl<T>;
+let createBreakCompletion: <T>(init: BreakCompletionInit<T>) => BreakCompletion<T>;
+let createContinueCompletion: <T>(init: ContinueCompletionInit<T>) => ContinueCompletion<T>;
+let createReturnCompletion: (init: ReturnCompletionInit) => ReturnCompletion;
+let createThrowCompletion: (init: ThrowCompletionInit) => ThrowCompletionImpl;
+
+type NormalCompletionInit<T> = Pick<NormalCompletion<T>, 'Type' | 'Value' | 'Target'>;
+
+type BreakCompletionInit<T> = Pick<BreakCompletion<T>, 'Type' | 'Value' | 'Target'>;
+
+type ContinueCompletionInit<T> = Pick<ContinueCompletion<T>, 'Type' | 'Value' | 'Target'>;
+
+type ReturnCompletionInit = Pick<ReturnCompletion, 'Type' | 'Value' | 'Target'>;
+
+type ThrowCompletionInit = Pick<ThrowCompletion, 'Type' | 'Value' | 'Target'>;
+
+type AbruptCompletionInit<T> =
+  | BreakCompletionInit<T>
+  | ContinueCompletionInit<T>
+  | ReturnCompletionInit
+  | ThrowCompletionInit;
+
+type CompletionInit<T> =
+  | NormalCompletionInit<T>
+  | AbruptCompletionInit<T>;
+
+@callable((_target, _thisArg, [completionRecord]) => {
   // 1. Assert: completionRecord is a Completion Record.
-  Assert(init instanceof Completion);
+  Assert(completionRecord instanceof Completion);
   // 2. Return completionRecord as the Completion Record of this abstract operation.
-  return init;
-}) // @ts-expect-error
-class Completion<T = unknown> {
-  constructor(init: CompletionRecord<T>) {
-    this.Type = init.Type;
-    this.Value = init.Value;
-    this.Target = init.Target;
-  }
+  return completionRecord;
+})
+class CompletionImpl<const T> {
+  declare readonly Type: 'normal' | 'break' | 'continue' | 'return' | 'throw';
+  readonly Value!: T | Value;
+  readonly Target!: JSStringValue | undefined;
 
-  readonly Type: CompletionType;
-  readonly Value: T;
-  readonly Target: string | undefined;
+  constructor(init: CompletionInit<T>) {
+    if (new.target === CompletionImpl) {
+      switch (init.Type) {
+        case 'normal':
+          return createNormalCompletion(init);
+        case 'break':
+          return createBreakCompletion(init);
+        case 'continue':
+          return createContinueCompletion(init);
+        case 'return':
+          return createReturnCompletion(init) as CompletionImpl<T>;
+        case 'throw':
+          return createThrowCompletion(init) as CompletionImpl<T>;
+        default:
+          throw new OutOfRange('new Completion', init);
+      }
+    }
+
+    const { Type, Value, Target } = init;
+    Assert(new.target.prototype.Type === Type);
+    this.Value = Value;
+    this.Target = Target;
+  }
 
   // NON-SPEC
   mark(m: GCMarker) {
     m(this.Value);
   }
+
+  static {
+    Object.defineProperty(this, 'name', { value: 'Completion' });
```

Comment:
I wonder why to define the name, does it change or is it used in an observable way?

Reply:
I can remove this. I had this named `CompletionImpl` temporarily while writing this PR and forgot to remove it after renaming back to `Completion`.

Reply:
Ah, never mind. The class name is still `CompletionImpl`, so this gives it the name `Completion` to match the export. Updating the name makes it easier to reason over when debugging or tracing with `console.log()`